File: utils/stream_filter.py
"""
utils/stream_filter.py — Filter out ads, trailers, tutorials, and files < 80MB.
"""

import re
from typing import List, Dict, Any

MIN_SIZE_BYTES = 80 * 1024 * 1024  # 80MB

REJECT_NAMES = [
    r"donation", r"donate", r"support\s+us", r"premium", r"upgrade",
    r"tutorial", r"how\s+to\s+download", r"how\s+to\s+watch",
    r"guide", r"readme", r"instructions",
    r"trailer", r"teaser", r"sample", r"preview", r"demo",
    r"advertisement", r"sponsor", r"promo",
    r"sign\s+in", r"sign\s+up", r"login", r"register",
    r"subscribe", r"membership", r"pricing",
    r"coming\s+soon", r"not\s+available", r"error",
    r"placeholder", r"loading", r"spinner",
]

REJECT_URLS = [
    r"/login", r"/signin", r"/register", r"/signup",
    r"/donation", r"/donate", r"/premium", r"/upgrade",
    r"/subscribe", r"/pricing",
    r"\.html$", r"/about", r"/contact", r"/privacy", r"/terms",
    r"/category/", r"/tag/", r"/page/", r"/feed",
    r"googleads", r"doubleclick", r"googlesyndication",
    r"facebook\.com", r"twitter\.com", r"instagram\.com",
    r"youtube\.com/watch", r"youtu\.be",
]

VIDEO_PATTERNS = [
    r"\.m3u8", r"\.mp4", r"\.mkv", r"\.ts\b", r"\.mov", r"\.avi", r"\.webm",
    r"/video/", r"/stream/", r"/hls/", r"/playlist", r"/cdn/",
    r"workers\.dev", r"pixeldrain", r"fsl-buckets", r"hubcloud",
    r"googleusercontent", r"hub\.latent", r"r2\.dev", r"savefiles",
    r"111477", r"cloudserver",
]


def parse_size(text: str) -> int:
    """Parse file size from text, returns bytes or 0 if not found."""
    if not text:
        return 0
    m = re.search(r"(\d+(?:\.\d+)?)\s*(GB|MB|TB|KB)", str(text), re.IGNORECASE)
    if not m:
        return 0
    val = float(m.group(1))
    unit = m.group(2).upper()
    if unit == "TB":
        return int(val * 1024**4)
    elif unit == "GB":
        return int(val * 1024**3)
    elif unit == "MB":
        return int(val * 1024**2)
    elif unit == "KB":
        return int(val * 1024)
    return 0


def should_reject(stream: Dict[str, Any]) -> bool:
    """Check if a stream should be rejected (ad, tutorial, small file, etc.)."""
    if not stream or not stream.get("url"):
        return True
    url = stream["url"]
    if not url.startswith("http"):
        return True

    # Reject torrents
    if stream.get("infoHash"):
        return True

    name = (stream.get("name") or "").lower()
    title = (stream.get("title") or "").lower()
    desc = (stream.get("description") or "").lower()
    url_lower = url.lower()
    bh = stream.get("behaviorHints") or {}
    filename = (bh.get("filename") or "").lower()

    # Check reject patterns
    for pattern in REJECT_NAMES:
        if re.search(pattern, name) or re.search(pattern, title):
            return True

    for pattern in REJECT_URLS:
        if re.search(pattern, url_lower):
            return True

    # Size check
    size = bh.get("videoSize") or 0
    if not size:
        size = parse_size(desc) or parse_size(filename) or parse_size(title)
    if size and size < MIN_SIZE_BYTES:
        return True

    # Must look like a video URL
    is_video = any(re.search(p, url_lower) for p in VIDEO_PATTERNS)
    if not is_video and size == 0:
        return True

    return False


def filter_streams(streams: List[Dict]) -> List[Dict]:
    """Filter out ads, trailers, tutorials, and small files."""
    if not streams:
        return []
    return [s for s in streams if not should_reject(s)]

File: src/utils/stream-filter.js
// src/utils/stream-filter.js — Filter ads, tutorials, small files (< 80MB)

const MIN_SIZE = 80 * 1024 * 1024; // 80MB

const REJECT_NAMES = [
  /donation/i, /donate/i, /support\s+us/i, /premium/i, /upgrade/i,
  /tutorial/i, /how\s+to\s+download/i, /how\s+to\s+watch/i,
  /guide/i, /readme/i, /instructions/i,
  /trailer/i, /teaser/i, /sample/i, /preview/i, /demo/i,
  /advertisement/i, /sponsor/i, /promo/i,
  /sign\s+in/i, /sign\s+up/i, /login/i, /register/i,
  /subscribe/i, /membership/i, /pricing/i,
  /coming\s+soon/i, /not\s+available/i, /error/i,
  /placeholder/i, /loading/i, /spinner/i,
];

const REJECT_URLS = [
  /\/login/i, /\/signin/i, /\/register/i, /\/signup/i,
  /\/donation/i, /\/donate/i, /\/premium/i, /\/upgrade/i,
  /\/subscribe/i, /\/pricing/i,
  /\.html$/i, /\/about/i, /\/contact/i, /\/privacy/i, /\/terms/i,
  /\/category\//i, /\/tag\//i, /\/page\//i, /\/feed/i,
  /googleads/i, /doubleclick/i, /googlesyndication/i,
  /facebook\.com/i, /twitter\.com/i, /instagram\.com/i,
  /youtube\.com\/watch/i, /youtu\.be/i,
];

function parseSize(text) {
  if (!text) return null;
  const m = String(text).match(/(\d+(?:\.\d+)?)\s*(GB|MB|TB|KB)/i);
  if (!m) return null;
  const v = parseFloat(m[1]);
  const u = m[2].toUpperCase();
  if (u === 'TB') return v * 1024 ** 4;
  if (u === 'GB') return v * 1024 ** 3;
  if (u === 'MB') return v * 1024 ** 2;
  if (u === 'KB') return v * 1024;
  return null;
}

export function shouldReject(s) {
  if (!s || !s.url || !s.url.startsWith('http')) return true;
  if (s.infoHash) return true; // No torrents
  const name = (s.name || '').toLowerCase();
  const title = (s.title || '').toLowerCase();
  const desc = (s.description || '').toLowerCase();
  const url = s.url.toLowerCase();
  const fn = (s.behaviorHints?.filename || '').toLowerCase();

  for (const p of REJECT_NAMES) {
    if (p.test(name) || p.test(title)) return true;
  }
  for (const p of REJECT_URLS) {
    if (p.test(url)) return true;
  }

  // Size check
  let size = s.behaviorHints?.videoSize || null;
  if (!size) size = parseSize(desc) || parseSize(fn) || parseSize(title);
  if (size !== null && size < MIN_SIZE) return true;

  // Must look like a video URL
  const videoPatterns = [
    /\.m3u8/i, /\.mp4/i, /\.mkv/i, /\.ts\b/i, /\.mov/i, /\.avi/i, /\.webm/i,
    /\/video\//i, /\/stream\//i, /\/hls\//i, /\/playlist/i, /\/cdn\//i,
    /workers\.dev/i, /pixeldrain/i, /fsl-buckets/i, /hubcloud/i,
    /googleusercontent/i, /hub\.latent/i, /r2\.dev/i, /savefiles/i,
    /111477/i, /cloudserver/i, /a\.111477/i,
  ];
  const isVideo = videoPatterns.some(p => p.test(url));
  if (!isVideo && size === null) return true;

  return false;
}

export function filterStreams(streams) {
  if (!Array.isArray(streams)) return [];
  return streams.filter(s => !shouldReject(s));
}

export { MIN_SIZE };
